File: code/omstretch.py
"""Independent cross-check of code/omstretch.c, and the meaning of rho.

`omstretch.c` computes rho(c) = the fewest runs needed to cover c rotation
classes inside a single OM-STRETCH.  This file rebuilds the same object from
the Python definitions used everywhere else in the repo -- sharing no code
with the C -- and checks the two agree at n = 5 and n = 6.

The reduction (see code/coset_lemma.py).  Inside an om-stretch every
transition is om, so all generators stay in one right coset of H = <a,b>,
|H| = (n-1)!, and that coset meets each rotation class exactly once.  In a
SPLIT-FREE walk every class is used at most once, so a stretch is exactly a
simple path in the right Cayley graph Cay(H; {a, b}):

    g -> g.a   continues the current run   (free)
    g -> g.b   ends it, starts a new one   (one run)

Runs cap at n-1 by themselves, because ord(a) = n-1 makes the (n-1)-st a-step
revisit the run's own start.  H acts on itself by left multiplication without
disturbing the right-multiplication edges, so the graph is vertex-transitive
and every path may be assumed to start at the identity.

Known values:

    n = 5   rho(24)  = 8      exact
    n = 6   rho(120) = 31     exact
    n = 7   rho(720) >= 128   partial

The n = 5 and n = 6 values are exact: the search proves every smaller run
count infeasible.  At n = 7 the same search certified K = 120 ... 127 all
impossible (the K = 127 level alone took 1.3e11 nodes) before the node cap
stopped it at K = 128, so rho(720) >= 128 is proved and the exact value is
open.
"""

import sys
from itertools import permutations

sys.path.insert(0, "code")

RHO = {5: 8, 6: 31}


def build(n):
    ident = tuple(range(1, n + 1))
    sig = lambda u: u[1:] + u[:1]
    delta = lambda u: u[2:] + (u[1], u[0])
    comp = lambda u, v: tuple(u[v[i] - 1] for i in range(n))
    a = ident
    for _ in range(n - 1):
        a = comp(a, sig(ident))
    a = comp(a, delta(ident))
    b = tuple(list(range(3, n)) + [2, 1, n])
    return ident, comp, a, b


def graph(n):
    """Vertices of H = <a,b>, with the two out-edges, indexed 0..(n-1)!-1."""
    ident, comp, a, b = build(n)
    idx, verts = {ident: 0}, [ident]
    head = 0
    while head < len(verts):
        x = verts[head]
        for g in (a, b):
            y = comp(x, g)
            if y not in idx:
                idx[y] = len(verts)
                verts.append(y)
        head += 1
    sa = [idx[comp(v, a)] for v in verts]
    sb = [idx[comp(v, b)] for v in verts]
    return len(verts), sa, sb


def loops(nv, sa, n):
    """a-cycles of the coset: the (n-2)! two-loops."""
    lid, pos = [-1] * nv, [0] * nv
    nl = 0
    for i in range(nv):
        if lid[i] >= 0:
            continue
        x, k = i, 0
        while True:
            lid[x], pos[x] = nl, k
            k += 1
            x = sa[x]
            if x == i:
                break
        assert k == n - 1, (k, n)
        nl += 1
    return nl, lid, pos


def arcs(mask, ln):
    """maximal circular arcs of set bits -- each needs its own run."""
    if mask == 0:
        return 0
    if mask == (1 << ln) - 1:
        return 1
    return sum(1 for i in range(ln)
               if (mask >> i & 1) and not (mask >> ((i - 1) % ln) & 1))


def rho_full(n, verbose=True):
    """Exact rho((n-1)!) by iterative deepening on the number of runs."""
    nv, sa, sb = graph(n)
    nl, lid, pos = loops(nv, sa, n)
    ln = n - 1
    table = [arcs(m, ln) for m in range(1 << ln)]

    for K in range(nl, nv + 1):
        vis = [False] * nv
        lmask = [(1 << ln) - 1] * nl
        state = {"arcsum": nl, "hit": False}

        def take(g):
            vis[g] = True
            L = lid[g]
            state["arcsum"] -= table[lmask[L]]
            lmask[L] &= ~(1 << pos[g])
            state["arcsum"] += table[lmask[L]]

        def give(g):
            L = lid[g]
            state["arcsum"] -= table[lmask[L]]
            lmask[L] |= 1 << pos[g]
            state["arcsum"] += table[lmask[L]]
            vis[g] = False

        def dfs(g, cov, runs, runlen):
            if cov == nv:
                state["hit"] = True
                return True
            if runs + state["arcsum"] - 1 > K:
                return False
            if runlen < ln:
                h = sa[g]
                if not vis[h]:
                    take(h)
                    if dfs(h, cov + 1, runs, runlen + 1):
                        return True
                    give(h)
            if runs < K:
                h = sb[g]
                if not vis[h]:
                    take(h)
                    if dfs(h, cov + 1, runs + 1, 1):
                        return True
                    give(h)
            return False

        take(0)
        ok = dfs(0, 1, 1, 1)
        if verbose:
            print(f"    K = {K:3}  {'FEASIBLE' if ok else 'impossible'}")
        if ok:
            return K
    return None


if __name__ == "__main__":
    print(__doc__.split("Known values:")[0].strip())
    for n in (5, 6):
        nv, sa, sb = graph(n)
        nl, _, _ = loops(nv, sa, n)
        import math
        print(f"\nn = {n}:  |H| = {nv} = (n-1)!, "
              f"{nl} loops = (n-2)!")
        assert nv == math.factorial(n - 1) and nl == math.factorial(n - 2)
        r = rho_full(n)
        print(f"  rho({nv}) = {r}   "
              f"{'MATCHES omstretch.c' if r == RHO[n] else 'MISMATCH!'}")
        assert r == RHO[n], (n, r, RHO[n])
    print("\nPython and C agree at n = 5 and n = 6.")

File: code/omstretch.c
/* omstretch.c -- the om-stretch primitive rho(c), exactly.
 *
 * Inside a single om-stretch every transition is om, so (Coset Lemma,
 * code/coset_lemma.py) all generators stay in one right coset of
 * H = <a,b>, |H| = (n-1)!, and that coset meets every rotation class
 * exactly once.  In a SPLIT-FREE walk each class is used at most once, so a
 * stretch is exactly a SIMPLE PATH in the right Cayley graph
 *
 *      Cay(H; {a, b}) on (n-1)! vertices, out-degree 2:
 *          g -> g.a   continues the current run   (free)
 *          g -> g.b   ends it and starts a new one (costs one run)
 *
 * Runs cap at n-1 automatically: ord(a) = n-1, so an (n-1)-st a-step would
 * revisit the run's own start.
 *
 *      rho(c) := min number of runs over simple paths covering c vertices
 *              =  1 + min number of b-edges.
 *
 * Left multiplication by h in H preserves right-Cayley edges, so the graph is
 * vertex-transitive and every path may be assumed to start at the identity.
 *
 * Build:  gcc -O3 -march=native -DNSYM=6 -o code/omstretch code/omstretch.c
 * Run:    ./code/omstretch [budget]
 *         with no budget it reports rho(c) for every c it can prove.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef NSYM
#define NSYM 6
#endif

#define N NSYM

static int NV;                 /* (n-1)!  vertices */
static int *succ_a, *succ_b;   /* the two out-edges */
static unsigned char *vis;
static int *bestrun;           /* bestrun[c] = min runs proven to cover c */
static long long nodes;
static long long node_cap;

/* ---- permutation helpers ------------------------------------------------ */
typedef struct { unsigned char p[N]; } perm;

static perm pcomp(perm u, perm v)            /* (u.v)[i] = u[v[i]] */
{
    perm w;
    for (int i = 0; i < N; i++) w.p[i] = u.p[v.p[i] - 1];
    return w;
}
static int peq(perm x, perm y) { return memcmp(x.p, y.p, N) == 0; }

static perm pident(void)
{
    perm u;
    for (int i = 0; i < N; i++) u.p[i] = i + 1;
    return u;
}
static perm psig(perm u)                     /* rotate left */
{
    perm w;
    for (int i = 0; i < N - 1; i++) w.p[i] = u.p[i + 1];
    w.p[N - 1] = u.p[0];
    return w;
}
static perm pdelta(perm u)
{
    perm w;
    for (int i = 0; i < N - 2; i++) w.p[i] = u.p[i + 2];
    w.p[N - 2] = u.p[1];
    w.p[N - 1] = u.p[0];
    return w;
}

/* ---- build a and b, then the coset graph -------------------------------- */
static perm A_, B_;

static void build_ab(void)
{
    perm id = pident(), c = psig(id), d = pdelta(id);
    perm a = id;
    for (int i = 0; i < N - 1; i++) a = pcomp(a, c);
    a = pcomp(a, d);                          /* a = c^(n-1) d, ord n-1 */
    A_ = a;

    /* b = the unique weight-(3) exit of a complete traversal whose cap is
     * n-1.  Rebuilding the cap test here would duplicate exit_table.py; the
     * element is characterised far more simply and is verified against the
     * Python side by omstretch.py:  b = (3,4,...,n-1,2,1,n).            */
    perm b;
    for (int i = 0; i < N - 3; i++) b.p[i] = i + 3;
    b.p[N - 3] = 2;
    b.p[N - 2] = 1;
    b.p[N - 1] = N;
    B_ = b;
}

static void build_graph(void)
{
    /* closure of <a,b> by BFS, indexing vertices in discovery order */
    int cap = 1;
    for (int i = 2; i <= N - 1; i++) cap *= i;      /* (n-1)! */
    NV = cap;

    perm *verts = malloc(sizeof(perm) * NV);
    succ_a = malloc(sizeof(int) * NV);
    succ_b = malloc(sizeof(int) * NV);

    int cnt = 0;
    verts[cnt++] = pident();
    for (int head = 0; head < cnt; head++) {
        perm x = verts[head];
        perm ys[2] = { pcomp(x, A_), pcomp(x, B_) };
        for (int k = 0; k < 2; k++) {
            int found = -1;
            for (int j = 0; j < cnt; j++)
                if (peq(verts[j], ys[k])) { found = j; break; }
            if (found < 0) {
                if (cnt >= NV) { fprintf(stderr, "closure exceeded (n-1)!\n"); exit(1); }
                verts[cnt] = ys[k];
                found = cnt++;
            }
            if (k == 0) succ_a[head] = found; else succ_b[head] = found;
        }
    }
    if (cnt != NV) { fprintf(stderr, "closure = %d, expected %d\n", cnt, NV); exit(1); }
    free(verts);
}

/* ---- loop decomposition, for the arc bound ------------------------------
 * The a-edges cut the coset into (n-2)! cycles of length n-1 (the 2-loops).
 * A run never leaves its loop, and the UNVISITED vertices of a loop fall into
 * maximal a-consecutive arcs -- each of which needs a run of its own.  So
 *
 *      remaining runs  >=  sum over loops of (#unvisited arcs)  -  1
 *
 * (the -1 because the run in progress may extend into one of them).  That is
 * far stronger than ceil(remaining / (n-1)).
 */
static int *loop_of, *pos_in_loop;
static int NL;                          /* (n-2)! loops */
static int *lmask;                      /* unvisited bitmask per loop */
static unsigned char arcs_of[1 << (N - 1)];
static int arcsum;                      /* sum of arcs_of[lmask[.]] */

static int count_arcs(int mask, int len)
{
    if (mask == 0) return 0;
    if (mask == (1 << len) - 1) return 1;          /* full cycle: one arc */
    int a = 0;
    for (int i = 0; i < len; i++) {
        int prev = (i + len - 1) % len;
        if ((mask >> i & 1) && !(mask >> prev & 1)) a++;
    }
    return a;
}

static void build_loops(void)
{
    loop_of = malloc(sizeof(int) * NV);
    pos_in_loop = malloc(sizeof(int) * NV);
    for (int i = 0; i < NV; i++) loop_of[i] = -1;
    NL = 0;
    for (int i = 0; i < NV; i++) {
        if (loop_of[i] >= 0) continue;
        int x = i, k = 0;
        do { loop_of[x] = NL; pos_in_loop[x] = k++; x = succ_a[x]; } while (x != i);
        if (k != N - 1) { fprintf(stderr, "a-cycle of length %d\n", k); exit(1); }
        NL++;
    }
    lmask = malloc(sizeof(int) * NL);
    for (int i = 0; i < NL; i++) lmask[i] = (1 << (N - 1)) - 1;
    for (int m = 0; m < (1 << (N - 1)); m++) arcs_of[m] = count_arcs(m, N - 1);
    arcsum = NL;                        /* every loop is one full arc */
}

static inline void take(int g)
{
    vis[g] = 1;
    int L = loop_of[g];
    arcsum -= arcs_of[lmask[L]];
    lmask[L] &= ~(1 << pos_in_loop[g]);
    arcsum += arcs_of[lmask[L]];
}
static inline void give(int g)
{
    int L = loop_of[g];
    arcsum -= arcs_of[lmask[L]];
    lmask[L] |= (1 << pos_in_loop[g]);
    arcsum += arcs_of[lmask[L]];
    vis[g] = 0;
}

/* ---- branch and bound --------------------------------------------------- */
static int target;             /* vertices we are trying to cover */
static int best;               /* best #runs found for `target` */

static void dfs(int g, int cov, int runs, int runlen)
{
    if (++nodes > node_cap) return;
    if (runs < bestrun[cov]) bestrun[cov] = runs;
    if (cov == target) { if (runs < best) best = runs; return; }

    /* Coverage bound.  The arc bound (every unvisited arc needs its own run)
     * is only valid when the path must cover EVERY vertex; for a partial
     * target the path may simply avoid the awkward arcs, so fall back to the
     * capacity bound there. */
    int freecap = (N - 1) - runlen;          /* the current run can still grow */
    int need = target - cov - freecap; if (need < 0) need = 0;
    int lo = runs + (need + N - 2) / (N - 1);
    if (target == NV) {
        int alt = runs + arcsum - 1;
        if (alt > lo) lo = alt;
    }
    if (lo >= best) return;

    if (runlen < N - 1) {
        int h = succ_a[g];
        if (!vis[h]) {
            take(h);
            dfs(h, cov + 1, runs, runlen + 1);
            give(h);
            if (nodes > node_cap) return;
        }
    }
    if (runs + 1 < best) {
        int h = succ_b[g];
        if (!vis[h]) {
            take(h);
            dfs(h, cov + 1, runs + 1, 1);
            give(h);
        }
    }
}

int main(int argc, char **argv)
{
    node_cap = (argc > 1) ? atoll(argv[1]) : 200000000LL;

    build_ab();
    build_graph();
    printf("n = %d   |<a,b>| = %d   (expected (n-1)! )\n", N, NV);
    printf("node cap = %lld\n", node_cap);

    vis = calloc(NV, 1);
    build_loops();
    bestrun = malloc(sizeof(int) * (NV + 1));
    for (int i = 0; i <= NV; i++) bestrun[i] = 1 << 20;
    printf("loops = %d (expected (n-2)!)\n", NL);

    /* iterative deepening on the number of runs: the first K that admits a
     * Hamiltonian path IS rho(NV), and each infeasible K is a proof. */
    target = (argc > 2) ? atoi(argv[2]) : NV;
    if (target < 1 || target > NV) { fprintf(stderr, "bad target\n"); exit(1); }
    int lb = (target == NV) ? NL : (target + N - 2) / (N - 1);
    int found = 0;
    for (int K = lb; K <= NV && !found; K++) {
        for (int i = 0; i < NV; i++) vis[i] = 0;
        for (int i = 0; i < NL; i++) lmask[i] = (1 << (N - 1)) - 1;
        arcsum = NL;
        best = K + 1;                   /* accept any solution with <= K runs */
        nodes = 0;
        take(0);
        dfs(0, 1, 1, 1);
        int capped = nodes > node_cap;
        printf("  c = %3d  K = %3d : %-12s nodes %lld%s\n", target, K,
               best <= K ? "FEASIBLE" : (capped ? "unknown" : "impossible"),
               nodes, capped ? "  (CAP HIT)" : "");
        if (best <= K) { found = 1; printf("rho(%d) = %d   (exact)\n", target, K); }
        else if (capped) { printf("cap hit at K = %d; rho(%d) > %d unproven\n", K, target, K - 1); break; }
    }

    printf("\n  c    rho(c) <=\n");
    for (int c = 1; c <= NV; c++)
        if (bestrun[c] < (1 << 20) &&
            (c % (N - 1) == 0 || c == NV || c <= 12))
            printf("%5d %9d\n", c, bestrun[c]);
    return 0;
}
